### PR TITLE
Authenticate SDS requests

### PR DESCRIPTION
**Pull request by katie-gardner**
### What is the context of this PR?
This PR updates launcher to authenticate SDS requests with GCP. It follows a similar approach to the previously merged [runner pr](https://github.com/ONSdigital/eq-questionnaire-runner/pull/1172)

The oauth2 library wasn't working with go version 1.16 so this PR has also updated to the latest stable version of go and updated packages.

### How to review
**Without GCP authentication**
Simply check that switching to this PR, without making any changes, still allows you to launch supplementary data surveys locally while the mock SDS is running

**GCP locally**
Similar to runner

1. If you already have a service account json file from testing the runner PR, you can use this - otherwise
   - Log into the GCP console in Chrome
   - Create or use an existing eq-sandbox project, add a service account and client id under APIs and Services > Credentials.
   - Download the service account json file and place it somewhere in the project.
2. Set the following environment variables for launcher
   - `SDS_OAUTH2_CLIENT_ID=<the Client Id in the GCP project>`
   - `GOOGLE_APPLICATION_CREDENTIALS=<path to the service account json>`
   - `OIDC_TOKEN_BACKEND=gcp`
   - If you're using a JetBrains tool, you can set these in your run configuration
   - If you're using VSCode, go to `.vscode/settings.json` and add the following
    ```
    {
        "go.testEnvVars": {
             "SDS_OAUTH2_CLIENT_ID": "<the Client Id in the GCP project>",
             "GOOGLE_APPLICATION_CREDENTIALS": "<path to the service account json>"
             "OIDC_TOKEN_BACKEND"="gcp"
         },
    }
    ```
    alternatively for `OIDC_TOKEN_BACKEND` you could change the default in `settings.go`
3. Open Launcher and pick `test_supplementary_data`
4. You should notice `Succesfully generated GCP ID token` in the logs
5. Refresh the page and pick the same schema, this time you should see `Found cached GCP ID token source`
6. To test the TTL Cache, you can change the OIDC_TOKEN_VALIDITY_IN_SECONDS and leeway in settings.go, or pass them in as environment variables, and pick a shorter time frame. Note that the cache is only cleared once per minute so if you want to test a shorter amount of time then change `time.Minute` in `oidc.go` to `time.Second` or similar

**GCP pipeline**
1. Follow the steps from the [PR supporting the mock SDS ](https://github.com/ONSdigital/eq-pipelines/pull/503) to create your own staging environment.
2. runner.image_tag can just be latest as the runner PR is merged
3. If you test prior to changing anything, following the steps on the PR should show launcher not working, but runner supporting supplementary data via quicklaunch
3. Then set the launcher image tag to `authenticate-with-sds` and update the pipeline
4. Launcher and runner should both now work
5. Note that you don't need to do anything with oidc_token_backend in the pipelines as this is set in the terraform repo and defaults to gcp

### CODE REVIEW

**Comment by katie-gardner** [comment]
[Launcher: Authenticate with SDS using Google ID Token (3 SP)](https://trello.com/c/g5rOv5xR/5801-launcher-authenticate-with-sds-using-google-id-token-3-sp)

File: settings/settings.go [hunk 1]
```diff
@@ -23,6 +23,10 @@ func init() {
 	setSetting("SDS_API_BASE_URL", "http://localhost:5003")
 	setSetting("JWT_ENCRYPTION_KEY_PATH", "jwt-test-keys/sdc-user-authentication-encryption-sr-public-key.pem")
 	setSetting("JWT_SIGNING_KEY_PATH", "jwt-test-keys/sdc-user-authentication-signing-launcher-private-key.pem")
+	setSetting("OIDC_TOKEN_VALIDITY_IN_SECONDS", "3600")
+	setSetting("OIDC_TOKEN_LEEWAY_IN_SECONDS", "300")
+	setSetting("OIDC_TOKEN_BACKEND", "local")
+	setSetting("SDS_OAUTH2_CLIENT_ID", "")
```
Comment (MebinAbraham) [inline settings/settings.go:29]: Add these to the `# Settings` section of the readme

File: oidc/oidc.go [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+			return nil, err
+		}
+		ttlCache.Set(audience, tokenSource, cache.DefaultExpiration)
+		return tokenSource, nil
+	}
+	return cachedFunc
+}
+
+// uses the Google Cloud metadata server environment to create an identity token that can be added to a HTTP request
+func getIdTokenFromMetadataServer(audience string) (oauth2.TokenSource, error) {
```
Comment (MebinAbraham) [inline oidc/oidc.go:52]: Since this is pretty much lifted from the google docs, worth linking to the source.

File: oidc/oidc.go [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+	if err != nil {
+		return nil, fmt.Errorf("failed to create NewTokenSource: %w", err)
+	}
+
+	// Generate the ID token.
+	_, err = ts.Token()
+	if err != nil {
+		return nil, fmt.Errorf("failed to receive token: %w", err)
+	}
+	log.Print("Succesfully generated GCP ID token")
```
Comment (MebinAbraham) [inline oidc/oidc.go:70]: Worth logging out the audience.

**Review by MebinAbraham — APPROVED** [review]
Looks good. Had to delete the `vendor` folder to be able to run locally. Also working with deployed version inc. caching. 👍

**Comment by katie-gardner** [comment]
> Looks good. Had to delete the `vendor` folder to be able to run locally. Also working with deployed version inc. caching. 👍

The `vendor` folder? don't have one of those 🤔

**Comment by MebinAbraham** [comment]
> > Looks good. Had to delete the `vendor` folder to be able to run locally. Also working with deployed version inc. caching. 👍
> 
> The `vendor` folder? don't have one of those 🤔

Probably historic.
```
 To ignore the vendor directory, use -mod=readonly or -mod=mod.
        To sync the vendor directory, run:
                go mod vendor
```

there is reference to the vendor in the lint.sh/fmt.sh files, which currently errors on run.

**Review by petechd — APPROVED** [review]
Works as expected in GCP!

**Pull request merged**